### PR TITLE
Seed raw data tables

### PR DESCRIPTION
**Pull request by rtownsend722**
- Define knex schemas for postgreSQL database
- Define Bookshelf models for queries
- Implement seed programs for raw data tables (users, tweets, favorites, networks)

### CODE REVIEW

File: db/knexfile.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ module.exports = {
     host: 'localhost',
     user: 'reb',
     password: '',
```
Comment (aploftus) [inline db/knexfile.js]: Consider setting a password, even something simple, to minimize risk of accidental modifications

File: db/migrations/20171214122219_setup.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+exports.up = function(knex, Promise) {
+  return Promise.all([
+
+    knex.schema.createTableIfNotExists('users', function(table) {
```
Comment (aploftus) [inline db/migrations/20171214122219_setup.js]: Did we agree to use ES6 function instantiation where possible?

If so, we should all add that rule to our linter files. I can track down the rule and push it out.

If not, ignore this.

File: db/migrations/20171214122219_setup.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  ])
+};
+
+exports.down = function(knex, Promise) {
+  return Promise.all([
+    knex.schema.dropTable('dailyengagements'),
+    knex.schema.dropTable('networkmetrics'),
+    knex.schema.dropTable('networks'),
+    knex.schema.dropTable('hourlyengagements'),
+    // knex.schema.dropTable('tweets'),
```
Comment (aploftus) [inline db/migrations/20171214122219_setup.js]: Remove unnecessary comments

File: db/migrations/20171216131810_step1.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+exports.up = function(knex, Promise) {
+  return Promise.all([
+
+    knex.schema.createTableIfNotExists('tweets', function(table) {
```
Comment (aploftus) [inline db/migrations/20171216131810_step1.js]: Do you need to save and/or commit all of these migrations files for kinex to work?

File: server/seedFavorite.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+/*====================GOAL: 25,000===================*/
+/*=================BIAS CONSTRAINTS=====================*
+ * even numbered (ie: non-influencer) users like more tweets than odd numbered (ie: influencer) users
+ ******* bias constrains below are for stress testing consideration *******
+ * 12% of likes belong to regular users (ie: other even ids)
+ * 25% go to bots
+ * 13% go to odd ids
+*========================================================*/
+
+let favorite = ``;
```
Comment (aploftus) [inline server/seedFavorite.js]: Use backticks only if string contains template literals

File: server/seedFavorite.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
+const faker = require('faker');
+const fs = require('fs');
+const randomNumber = require('./helpers.js').randomNumber;
+const outfile = '../server/blahblah.csv';
```
Comment (aploftus) [inline server/seedFavorite.js]: Consider renaming your csv file to something more descriptive

File: server/seedFavorite.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ * 12% of likes belong to regular users (ie: other even ids)
+ * 25% go to bots
+ * 13% go to odd ids
+*========================================================*/
+
+let favorite = ``;
+
+// Non-Influencers
+const generateRegularFavorites = () => {
+  for (var i = 1; i < 500001; i++) {
```
Comment (aploftus) [inline server/seedFavorite.js]: Might be useful to include comments detailing what each large number or for loop is counting up to. Users, tweets, etc

File: server/seedFavorite.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    if (randomNumber(100) < 13) {
+      for (let j = 0; j < 20; j++) {
+        favorite = `${faker.date.recent().toISOString()}, ${faker.date.recent().toISOString()}, ${randomNumber(10000000) + 1}, ${i}\r\n`;
```
Comment (aploftus) [inline server/seedFavorite.js]: Also helpful to detail what field each of these variables will be filling

File: server/seedFavorite.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        fs.appendFileSync(outfile, favorite, err => {
+          console.log(err);
+        }); 
+      }
+    }
+  }
+}
+
+const generateInfluencerFavorites = () => {
+  for (var i = 2; i < 50001; i+=2) {
```
Comment (aploftus) [inline server/seedFavorite.js]: Here you increment by 2, to get even i's. Above, you increment by 1. Is this intentional?

Comment (aploftus) [reply server/seedFavorite.js]: Additionally, your comments above say that even numbered users are the non-influencers.

File: server/seedNetwork.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+let network = '';
+let date;
+
+//25,000 influencers
+const generateInfluencerFollows = () => {
+  for (let i = 1; i < 50001; i+=2) {
+    date = faker.date.recent().toISOString();
+    for (let j = 0; j < 10; j++) {
+      network = `${date}, ${date}, ${i}, ${randomNumber(50000) + 1}\r\n`;
```
Comment (aploftus) [inline server/seedNetwork.js]: Here the numbers are easier to follow. Also might be helpful to detail what field each of these variables will be filling

File: server/seedTweet.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+const makeTweet = (preface, noun, word) => {
+  return `${preface} ${noun} are ${word}`;
+};
+
+let tweet = '';
+
+const outputTweets = () => {
+  for (let i = 1; i < 10000001; i++) {
+    let user_id = randomNumber(50000) + 1;
+    let prefaceIndex = randomNumber(4);
```
Comment (aploftus) [inline server/seedTweet.js]: You might consider putting the math / index-picking in your makeTweet to have that method then be a makeRandomTweet, more modular. This makes sense, though.

File: server/seedTweet.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  'chatty cashiers',
+  'my imaginary friends from elementary school'
+];
+
+
+const makeTweet = (preface, noun, word) => {
+  return `${preface} ${noun} are ${word}`;
+};
+
+let tweet = '';
```
Comment (aploftus) [inline server/seedTweet.js]: I would move this inside outputTweets but outside the for loop, to take it off global scope. You could also more the user_id, prefaceIndex, etc Declarations outside the for loop, and just reassign in the for loop, to save up on memory as well.

File: server/seedTweet.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+      console.log(err);
+    });
+  }
+}
+/*===========UNCOMMENT TO GENERATE TWEETS===========*/
+outputTweets();
+
+
+/*=========EFFICIENCY LOG=========*/
+// Times out at 90167 calls
```
Comment (aploftus) [inline server/seedTweet.js]: is this still the case?

File: server/seedTweet.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    } else {
+      tweet = `${i}, ${user_id}, ${makeTweet(prefaces[prefaceIndex], nouns[nounIndex], neutralWords[wordIndex])}, 0\r\n`;
+    }
+    fs.appendFileSync(outfile, tweet, err => {
+      console.log(err);
+    });
+  }
+}
+/*===========UNCOMMENT TO GENERATE TWEETS===========*/
+outputTweets();
```
Comment (aploftus) [inline server/seedTweet.js]: Do you want this commented out, (as you have done in the other files), so you don't accidentally build 10M tweets?

File: server/seedUser.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
+const underscore = require('underscore');
```
Comment (aploftus) [inline server/seedUser.js]: I don't think you're using this dependency

File: server/routes.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
+const routes = require('express').Router();
+const _ = require('underscore');
+const Promise = require('bluebird');
+const User = require('../db/user.js');
+const users = require('./seedUser.js').users;
+
+/*==========UNCOMMENT TO SEED USERS==========*/
+// routes.get('/seedUsers', (req, res) => {
```
Comment (aploftus) [inline server/routes.js]: Is there a way to import your users not through an http request, but rather the same way you are importing tweets and favorites? (CSV?)

File: server/seedUser.js (hunk 15 above)
Comment (aploftus) [reply server/seedUser.js]: I'm interested in why the user generation is done differently than all the other historical data generation.

File: db/dailyengagement.js [hunk 17]
```diff
@@ -0,0 +1,7 @@
+const Bookshelf = require('./bookshelf.js');
+
+const Dailyengagement = Bookshelf.Model.extend({
+  tableName: 'dailyengagements'
```
Comment (aploftus) [inline db/dailyengagement.js]: What fields/columns will be in this table?

File: server/seedFavorite.js (hunk 6 above)
Comment (rtownsend722) [reply server/seedFavorite.js]: lolololol good catch. I was using blahblah to patch some random mistakes!

**Comment by rtownsend722** [comment]
Thanks for all the great feedback! I implement the majority. ES6 with Knex is unstable, so I'm using ES5 for those models. Bookshelf works in conjunction with Knex - so you define your fields in the Knex migration files, and not in the Bookshelf models. Good q. re: populating users via HTTP. Before using the export to CSV method, I was attempting to bulk upload via HTTP, and just forgot to remove the route!

**Pull request closed without merging**